x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/profile/TransportActivateProfileAction.java

```diff
@@ -0,0 +1,85 @@
+
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.action.GrantApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileAction;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileRequest;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileResponse;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.support.UsernamePasswordToken;
+import org.elasticsearch.xpack.security.authc.AuthenticationService;
+import org.elasticsearch.xpack.security.authc.BearerToken;
+import org.elasticsearch.xpack.security.profile.ProfileService;
+
+public class TransportActivateProfileAction extends HandledTransportAction<ActivateProfileRequest, ActivateProfileResponse> {
+
+    private final ProfileService profileService;
+    private final AuthenticationService authenticationService;
+    private final ThreadContext threadContext;
+
+    @Inject
+    public TransportActivateProfileAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        ProfileService profileService,
+        AuthenticationService authenticationService,
+        ThreadPool threadPool
+    ) {
+        super(ActivateProfileAction.NAME, transportService, actionFilters, ActivateProfileRequest::new);
+        this.profileService = profileService;
+        this.authenticationService = authenticationService;
+        this.threadContext = threadPool.getThreadContext();
+    }
+
+    @Override
+    protected void doExecute(Task task, ActivateProfileRequest request, ActionListener<ActivateProfileResponse> listener) {
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            resolveAuthentication(
+                request.getGrant(),
+                request,
+                ActionListener.wrap(
+                    authentication -> profileService.activateProfile(authentication, listener.map(ActivateProfileResponse::new)),
+                    listener::onFailure
+                )
+            );
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    // TODO: extract
+    private void resolveAuthentication(
+        GrantApiKeyRequest.Grant grant,
+        TransportRequest transportRequest,
+        ActionListener<Authentication> listener
+    ) {
+        switch (grant.getType()) {
```

Comment:
The switch basically chooses between `new UsernamePasswordToken(grant.getUsername(), grant.getPassword())` and `new BearerToken(grant.getAccessToken())`, plus it validates the `grant.Type()` request parameter. I think this method can be improved for readability.

Reply:
This method is also copied from `TransportGrantApiKeyAction`. I had a `TODO` in the beginning of this method and was meant to refactor it in a separate PR. But we may as well handle it now. I extracted `Grant` into a separate class and extract base classes for both GrantApiKey and ActivateProfile related classes. Hopefully the new change makes it read better.